format.swift

```diff
@@ -0,0 +1,548 @@
+    )
+  }
+
+  private func addHandler(
+    pipe: Pipe,
+    group: DispatchGroup,
+    addData: @escaping (Data) -> Void
+  ) {
+    group.enter()
+    pipe.fileHandleForReading.readabilityHandler = { fileHandle in
+      // Apparently using availableData can cause various issues
+      let newData = fileHandle.readData(ofLength: Int.max)
+      if newData.count == 0 {
+        pipe.fileHandleForReading.readabilityHandler = nil;
+        group.leave()
+      } else {
+        addData(newData)
+      }
+    }
+  }
+}
+
+/// The exit code and output (if redirected) from a subprocess that has
+/// terminated
+struct ProcessResult {
+  let stdout: String
+  let stderr: String
+}
+
+/// Error thrown if a process terminates with a non-zero exit code.
+struct NonZeroExitCodeError: Error {
+  let stdout: String
+  let stderr: String
+  let exitCode: Int
+}
+
+/// Error thrown if `stdout` or `stderr` could not be decoded as UTF-8.
+struct FailedToDecodeUTF8Error: Error {
+  let data: Data
+}
+
+// MARK: Actual implementation
+
+/// Primitive command line argument parsing for this script.
+///
+/// Since this script isn’t a package, we can’t depend on swift-argument-parser,
+/// which is a little unfortunate.
+enum CommandLineArguments {
+  static var help: Bool {
+    hasFlag("--help")
+  }
+
+  /// Whether to lint instead of format the files.
+  static var lint: Bool {
+    hasFlag("--lint")
+  }
+
+  /// Whether to update the swift-format that is checked out.
+  static var update: Bool {
+    hasFlag("--update")
+  }
+
+  /// The argument of the `--swift-format` argument.
+  static var swiftFormat: String? {
+    let arguments = ProcessInfo.processInfo.arguments
+    guard let labelIndex = arguments.firstIndex(of: "--swift-format") else {
+      return nil
+    }
+    let valueIndex = arguments.index(after: labelIndex)
+    if valueIndex == arguments.endIndex {
+      return nil
+    }
+    return arguments[valueIndex]
+  }
+
+  /// Whether to show verbose output of the steps that are being performed.
+  static var verbose: Bool {
+    hasFlag("--verbose")
+  }
+
+  private static func hasFlag(_ flag: String) -> Bool {
+    ProcessInfo.processInfo.arguments.contains(flag)
+  }
+}
+
+struct SwiftSyntaxFormatter {
+  enum Error: Swift.Error, CustomStringConvertible {
+    case swiftFormatNotFound
+    case lintFailed
+
+    var description: String {
+      switch self {
+      case .swiftFormatNotFound:
+        return "The locally built swift-format could not be found"
+      case .lintFailed:
+        return """
+          The swift-syntax repo is not formatted according to the style guides.
+
+          Run the following command to format swift-syntax
+
+            swift-syntax/format.swift
+
+          If the issue persists, try updating swift-format by running
+
+            swift-syntax/format.swift --update
+
+          Should that still fail, fix any remaining issues manually and verify they match the swift-format style using
```

Comment:
Do we expect this to fail at all?

Reply:
During the initial introduction of swift-format, I had two or three cases where swift-format complained but wasn’t able to fix the issue by itself. I don’t fee super strongly about this hint.

Reply:
Huh, I didn't know that was a thing. Feel free to keep, is just an extra sentence 👍